### ISSUE TITLE
Calling `failmsg` and `fail` msi via the PREP fails with `RE_TYPE_ERROR`

### ISSUE THREAD

**Issue by mstfdkmn** [irods/irods#7096]
### iRODS Version, OS and Version
iRODS 4.2.11/4.3.0 centos7/almalinux8

### What did you try to do?
I tried to call `failmsg` via callback in the Python Rule Engine Plugin.

### Expected behavior
Not to see any errors, and expect the action that calls the `failmsg` fails.

### Observed behavior (including steps to reproduce, if applicable)

I wanted to fire the PEP below by adding a metadata. I expect it to fail with errors "-169000 SYS_NOT_ALLOWED" in the client side.
Also I wanted to see the ADMINISTRATIVE_METADATA message in the logs.

```
def pep_api_mod_avu_metadata_pre(rule_args, callback, rei):
    callback.failmsg(-169000, "ADMINISTRATIVE_METADATA")
    #callback.error("SYS_NOT_ALLOWED")
    #callback.msiOprDisallowed()
```

But I get `-1230000 RE_TYPE_ERROR`.

```
[irods@dev-u0137480 irods]$ imeta add -d test.txt atr val
remote addresses: 10.0.2.100 ERROR: Level 0: DEBUG: simplify: unsolvable typing constraint STRING * STRING < INTEGER * STRING.

line 0, col 7
failmsg(*ARG0,*ARG1)
       ^


remote addresses: 10.0.2.100 ERROR: rcModAVUMetadata failed with error -1230000 RE_TYPE_ERROR
```

logs:
```
irods stdout | May  5 16:02:52 pid:3208 DEBUG: simplify: unsolvable typing constraint STRING * STRING < INTEGER * STRING.
irods stdout | line 0, col 7
irods stdout | failmsg(*ARG0,*ARG1)
irods stdout |        ^
irods stdout | {"log_category":"legacy","log_facility":"local0","log_level":"error","log_message":"simplify: unsolvable typing constraint STRING * STRING < INTEGER * STRING.\n\nline 0, col 7\nfailmsg(*ARG0,*ARG1)\n       ^\n\n","request_api_name":"MOD_AVU_METADATA_AN","request_api_number":706,"request_api_version":"d","request_client_user":"rods","request_host":"10.0.2.100","request_proxy_user":"rods","request_release_version":"rods4.3.0","server_host":"dev-u0137480.irods.t.icts.kuleuven.be","server_pid":3208,"server_timestamp":"2023-05-05T14:02:52.882Z","server_type":"agent"}
irods stdout | {"log_category":"legacy","log_facility":"local0","log_level":"error","log_message":"caught python exception: Traceback (most recent call last):\n  File \"/etc/irods/core.py\", line 52, in pep_api_mod_avu_metadata_pre\n    callback.failmsg(-169000, \"ADMINISTRATIVE_METADATA\")\nRuntimeError: [iRods__Error__Code:-1230000] [-]\t/irods_source/server/re/include/irods/irods_re_plugin.hpp:333:irods::error irods::dynamic_operation_execution_manager<std::tuple<>, RuleExecInfo *, irods::DONT_AUDIT_RULE>::call(const std::string &, const std::string &, OP, As &&...) [T = std::tuple<>, C = RuleExecInfo *, Audit = irods::DONT_AUDIT_RULE, OP = std::function<irods::error (const std::string &, irods::re_pack_inp<std::tuple<>> &, irods::unpack &&)>, As = <const std::string &, irods::re_pack_inp<std::tuple<>> &, irods::unpack>] :  status [RE_TYPE_ERROR]  errno [] -- message [applyRuleUpdateParams failed for rule failmsg]\n\t[-]\t/irods_source/plugins/rule_engines/irods_rule_language/src/irods_rule_language.cpp:324:irods::error exec_rule(irods::default_re_ctx &, const std::string &, std::list<boost::any> &, irods::callback) :  status [RE_TYPE_ERROR]  errno [] -- message [applyRuleUpdateParams failed for rule failmsg]\n\n\n","request_api_name":"MOD_AVU_METADATA_AN","request_api_number":706,"request_api_version":"d","request_client_user":"rods","request_host":"10.0.2.100","request_proxy_user":"rods","request_release_version":"rods4.3.0","server_host":"dev-u0137480.irods.t.icts.kuleuven.be","server_pid":3208,"server_timestamp":"2023-05-05T14:02:52.885Z","server_type":"agent"}
```

Note: These msi(s) work well in the native rule engine. Also, other msi like `writeLine` or `msiOprDisallowed` have no issue in the PREP.

**Comment by alanking** [comment]
I'm not sure if it works, but the `failmsg` microservice is used by the core.py template packaged with the PREP:
https://github.com/irods/irods_rule_engine_plugin_python/blob/5bf88f4cb8211828038d1d5746dbb571bd13e57c/core.py.template#L103 Have you played around with passing a string instead of an integer?

Although, it seems that PREP is serializing the parameter as a string and the microservice is expecting an integer, so that may not work, either.

Thoughts, @d-w-moore?

**Comment by d-w-moore** [comment]
As far as I know, fail and failmsg (and perhaps other microserviceds) were designed to work in the context of the NREP, and not necessarily outside of it.  failing in the PREP is generally handled via exception handling https://slides.com/danielmoore-5/2020-ku_leuven-irods-training-python-rule-engine#/11

**Comment by d-w-moore** [comment]
> Is that documented in the PREP repo? If not, we should add a section stating that.
I agree,  although I think this may be as much a framework thing   as a PREP thing. Eg how are failed rule calls supposed to work cross- language?

**Comment by d-w-moore** [comment]
If we were committing to doing this generically, I think it'd still depend on the facilities offered by XYZ language . (Substitute whatever  is the name of the language for which we're writing the rule language plugin for "XYZ"). 
   - if a rule in XYZ language were calling into a native rule a failing rule should(?) either return a negative value (error code) or raise an exception according to the conventions of XYZ.  however, NREP  functions ie rules "return" a value other than the error code (which is supposed to be contained in the fail mechanism instead of the function return mechanism), unlike Python, which returns the error code meant to indicate fail/not-fail/continue.
   - if native rules are calling into XYZ rules, how do we standardize? we'd probably require a fail in language-XYZ to be indicated by raising an error/exception in whatever is the typical XYZ-language fashion.  If we return error codes through the function-return mechanism of the language itself, how exactly should the calling language interpret it?

Or ... maybe we don't have to worry about being generic.  But we ultimately should endeavor to make things consistent for folks who want to write language plugins.  Esp. if, say, Go, Val, or Rust takes off and antiquates Python and/or C++.  

Maybe better to defer to @korydraughn for next steps / thoughts in this matter, if this strikes enough people as being a valid concern.

**Comment by d-w-moore** [comment]
@mstfdkmn  What happens if you simply return -169000 from 

Considering @alanking's point above, that an `ac*` rule in **core.py** is able to leverage `failmsg`,  possibly there is something in the processing of new-style PEPs (as opposed to the old-style action rules of the `ac*` name pattern) that blocks us from using `failmsg` in a python rule.

**Comment by mstfdkmn** [comment]
> I'm not sure if it works, but the `failmsg` microservice is used by the core.py template packaged with the PREP: https://github.com/irods/irods_rule_engine_plugin_python/blob/5bf88f4cb8211828038d1d5746dbb571bd13e57c/core.py.template#L103 Have you played around with passing a string instead of an integer?
> 
Yes, I found that line in the core.py template to be sure whether there is any example that shows how to call this msi in a rule...I did pass string instead of an integer and I got the exact same errors.

**Comment by mstfdkmn** [comment]
> @mstfdkmn What happens if you simply return -169000 from `pep_api_mod_avu_metadata_pre`?
> 

Getting the runtime error.

core.py:

```
def pep_api_mod_avu_metadata_pre(rule_args, callback, rei):
    """
    Triggered by
    - iCommands' imeta command
    - PRC MetadataManager

    Defined to block removing request of administrative metadata (mg. prefixes)
    """
    user_name = session_vars.get_map(rei)["client_user"]["user_name"]
    items = dict(rule_args[2].items())
    if items["arg0"] in ("rm", "rmw") and items["arg1"] == "-C" and items["arg3"].startswith("mg.") and user_name not in ("rods", "operator"):
        #callback.failmsg("-169000", "ADMINISTRATIVE_METADATA")
        #callback.msiOprDisallowed()
        #callback.error("SYS_NOT_ALLOWED")
        raise RuntimeError(-169000)
```

client:
```
[mustafa@dev-u0137480 ~]$ imeta rm -C /u0137480/home/mygroup mg.collection_type project_root
remote addresses: 10.0.2.100 ERROR: rcModAVUMetadata failed with error -1 Unknown iRODS error Operation not permitted

```
logs:
`irods stdout | {"log_category":"legacy","log_facility":"local0","log_level":"error","log_message":"caught python exception: Traceback (most recent call last):\n  File \"/etc/irods/core.py\", line 77, in pep_api_mod_avu_metadata_pre\n    raise RuntimeError(-169000)\nRuntimeError: -169000\n","request_api_name":"MOD_AVU_METADATA_AN","request_api_number":706,"request_api_version":"d","request_client_user":"mustafa","request_host":"10.0.2.100","request_proxy_user":"mustafa","request_release_version":"rods4.3.0","server_host":"dev-u0137480.irods.t.icts.kuleuven.be","server_pid":78527,"server_timestamp":"2023-05-09T07:33:13.932Z","server_type":"agent"}`

Also fyi; calling the rule in the slide you posted via irule results in `-1211000 RE_UNABLE_TO_READ_LOCAL_VAR`. So I could not manage to test  cross-language call of the python exception/erros.

**Comment by d-w-moore** [comment]
> Yes, I found that line in the core.py template to be sure whether there is any example that shows how to call this msi in a rule...I did pass string instead of an integer and I got the exact same errors.

I'd expect that; Python parameters, in- or outgoing, may be strings or integers, as well as a limited list of other types (when calling microservices for example).

Thanks, @mstfdkmn , for the tip -- the command line argument in that slide should've been with an asterisk -- `"*x=5"`

**Comment by d-w-moore** [comment]
@mstfdkmn  - we'll check into how to fail properly from a Python dynamic PEP.

**Comment by korydraughn** [comment]
It's likely that `failmsg` and other microservices are only meant to be used from the NREP.

However, there's no reason we can't implement microservices that support the following:
- Callable from any REP
- Adds an error code and message to the rError stack (for the client)
    - e.g. `msi_rError_append(code:int, msg:string)`
- Logs a message at a specific log level
    - e.g. `msi_log(log_category:string, log_level:string, msg:string)`

Calling these microservices would not cause any rule to fail. The policy implementer would still need to follow the recommendations documented by the REP. But, they would at least have a flexible way to achieve the wanted behavior.

**Comment by d-w-moore** [comment]
Hello , @mstfdkmn - following Kory's direction above, you could now use:
  - https://github.com/d-w-moore/rerr_msvc
  - https://github.com/d-w-moore/log_msvc
To effect a failure from a Python  rule.

Below is a possible application, similar to your original example (and using iRODS 4.3.0):
```
# /etc/irods/core.py

import irods_errors

_reverse_error_lookup = {val:key for key,val in vars(irods_errors).items() if type(val) is int}

def error_code_with_message(Code, Msg, callback, Level = "Info"):
     callback.msi_rError_append(abs(Code), Msg)
     if Code < 0:
         Msg = str(_reverse_error_lookup[Code]) + ": " + Msg
     callback.msi_log(Level, Msg)
     return Code

def pep_api_mod_avu_metadata_pre(rule_args,callback,e):
    return error_code_with_message(irods_errors.BAD_FUNCTION_CALL, # <-- or some negative integer code
                                   "something bogus is afoot!", callback)
```
An AVU modifying command eg `imeta -C . a b` should then show the client message with appropriate error code:
```
remote addresses: 127.0.0.1 ERROR: Level 0: something bogus is afoot!
remote addresses: 127.0.0.1 ERROR: rcModAVUMetadata failed with error -1823000 BAD_FUNCTION_CALL 
``` 
In addition,  a log message should be generated for the error.  Here's what I got in `/var/log/irods/irods.log`:
```
{
  "log_category": "legacy",
  "log_facility": "local0",
  "log_level": "info",
  "log_message": "BAD_FUNCTION_CALL: something bogus is afoot!",
  "request_api_name": "MOD_AVU_METADATA_AN",
  "request_api_number": 706,
  "request_api_version": "d",
  "request_client_user": "rods",
  "request_host": "127.0.0.1",
  "request_proxy_user": "rods",
  "request_release_version": "rods4.3.0",
  "server_host": "91d4cbfbc06b",
  "server_pid": 4393,
  "server_timestamp": "2023-05-16T16:07:19.358Z",
  "server_type": "agent"
}

```

**Comment by mstfdkmn** [comment]
Hi @d-w-moore  thanks for this. I would like to use this `error_code_with_message` function that you shared however I could not find easily any instruction about how to install and configure the microservices that you made. I guess we should make rpms and use/configure them as a cpp rule? If you could provide a guideline that would be great. Or could you refer me a build environment/Dockerfile etc?

**Comment by d-w-moore** [comment]
@mstfdkmn ,  You would have to build these microservices, as they are not official yet.  In a nutshell, for each microservice:
   - clone the github repo locally.
   - make a `build` directory, and cd into that directory.
   -  and invoking cmake from that build directory (first with the root repo as the argument, then either make or `cmake --build . --target package` to build the package).  Then install from the locally built package file.

Examples are provided in some of the iRODS training slides, eg. here: https://slides.com/irods/ugm2022-compute-to-data#/6

Since the microservices are (for the present case) 4.3.0 native, some adaptation of the source code in these repos  may be necessary if you are running on iRODS 4.2.X.

**Comment by mstfdkmn** [comment]
Hello @d-w-moore, I was able to build microservices and my test showed it worked as you described above. I am thinking to install them in our production environment however I wonder if I should expect these microservices will be officially released in 4.3.1. Thanks.

**Comment by mstfdkmn** [comment]
Sorry one last thing - where is the `irods_errors` -built in module- documented? I could find neither irods_rule_engine_plugin_python nor irods repository. I checked available irods error messages by `print(dir(irods_errors))` though.

**Comment by korydraughn** [comment]
> Hello @d-w-moore, I was able to build microservices and my test showed it worked as you described above. I am thinking to install them in our production environment however I wonder if I should expect these microservices will be officially released in 4.3.1. Thanks.

No official decision has been made on whether these MSIs will be included in a 4.3 release. When we know more, we'll be sure to let the community know.

> Sorry one last thing - where is the `irods_errors` -built in module- documented? I could find neither irods_rule_engine_plugin_python nor irods repository. I checked available irods error messages by `print(dir(irods_errors))` though.

@d-w-moore ping

**Comment by d-w-moore** [comment]
> Sorry one last thing - where is the `irods_errors` -built in module- documented? I could find neither in irods_rule_engine_plugin_python nor in the irods repository. I checked available irods error messages by `print(dir(irods_errors))` though.

It's docked in this section: 
https://github.com/irods/irods_rule_engine_plugin_python#python-globals-and-build-in-modules

**Comment by mstfdkmn** [comment]
> Hello , @mstfdkmn - following Kory's direction above, you could now use a combination of these microservices:
> 
> * https://github.com/d-w-moore/rerr_msvc
> * https://github.com/d-w-moore/log_msvc
> 
> To effect a failure from a Python rule.
> 
> Below is a possible application, similar to your original example (and using iRODS 4.3.0):
> 
> ```
> # /etc/irods/core.py
> 
> import irods_errors
> 
> _reverse_error_lookup = {val:key for key,val in vars(irods_errors).items() if type(val) is int}
> 
> def error_code_with_message(Code, Msg, callback, Level = "Info"):
>      callback.msi_rError_append(abs(Code), Msg)
>      if Code < 0:
>          Msg = str(_reverse_error_lookup[Code]) + ": " + Msg
>      callback.msi_log(Level, Msg)
>      return Code
> 
> def pep_api_mod_avu_metadata_pre(rule_args,callback,e):
>     return error_code_with_message(irods_errors.BAD_FUNCTION_CALL, # <-- or some negative integer code
>                                    "something bogus is afoot!", callback)
> ```
> 
> An AVU modifying command eg `imeta -C . a b` should then show the client message with appropriate error code:
> 
> ```
> remote addresses: 127.0.0.1 ERROR: Level 0: something bogus is afoot!
> remote addresses: 127.0.0.1 ERROR: rcModAVUMetadata failed with error -1823000 BAD_FUNCTION_CALL 
> ```
> 
> In addition, a log message should be generated for the error. Here's what I got in `/var/log/irods/irods.log`:
> 
> ```
> {
>   "log_category": "legacy",
>   "log_facility": "local0",
>   "log_level": "info",
>   "log_message": "BAD_FUNCTION_CALL: something bogus is afoot!",
>   "request_api_name": "MOD_AVU_METADATA_AN",
>   "request_api_number": 706,
>   "request_api_version": "d",
>   "request_client_user": "rods",
>   "request_host": "127.0.0.1",
>   "request_proxy_user": "rods",
>   "request_release_version": "rods4.3.0",
>   "server_host": "91d4cbfbc06b",
>   "server_pid": 4393,
>   "server_timestamp": "2023-05-16T16:07:19.358Z",
>   "server_type": "agent"
> }
> ```

Hi @d-w-moore, apparently these microservices cannot block actions of PEPs used but just throw informative messages both in the client side and in server logs. If this is expected then I probably misunderstood. I was guessing `irods_errors.BAD_FUNCTION_CALL` or another error code like `irods_errors."USER_NOT_ALLOWED_TO_EXEC_CMD` should stop the process. The only way for these MSIs to work correctly seems to use them in combination with a native one - `error`?

```
def pep_api_mod_avu_metadata_pre(rule_args,callback,e):
    error_code_with_message(irods_errors.BAD_FUNCTION_CALL, # <-- or some negative integer code
                                   "something bogus is afoot!", callback)
    callback.error("SYS_NOT_ALLOWED")

```

**Comment by d-w-moore** [comment]
@mstfdkmn - You need to return the (negative integer) error code from the `*_pre` PEP. 

The original example contained something more of this style
```
def pep_api_mod_avu_metadata_pre(rule_args,callback,rei):
    return error_code_with_message(irods_errors.SYS_NOT_ALLOWED, 
                                   "something bogus is afoot!", callback)
```
Notice the "return" there, which  made sure an error code actually gets returned from the Python rule.
If  the the `pep_<OPERATION>_pre`  PEP does not contain an explicit return statement, to return the negative integer code of the indicated iRODS error to the rule framework, then effectively zero (aka a success code) will be returned, causing the <OPERATION> itself to go through as normal.
